sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala

```diff
@@ -352,10 +418,7 @@ class StateStoreChangelogReaderV2(
 
   override def version: Short = 2
 
-  // ensure that the version read is v2
-  val changelogVersionStr = input.readUTF()
-  assert(changelogVersionStr == "v2",
-    s"Changelog version mismatch: $changelogVersionStr != v2")
+  verifyVersion()
```

Comment:
Can you have this be called in `StateStoreChangelogReader` initialization? I'd have the following in `StateStoreChangelogReader`
```scala
protected def readVersion(): String = input.readUTF()

final def verifyVersion(): Unit = {
  if(!readVersion.stripPrefix("v") == version.toString) {
    throw ...
  }
}

verifyVersion()
 ...
}

class StateStoreChangelogReaderV1 {
  override protected def readVersion(): String = "v1"
}
```

Reply:
Also if `readVersion` fails, then you also probably want to throw the "Upgrade Spark" error

Reply:
Yes this is the ideal way, but the problem is in v1 there is no version written in the first line. Hence all the problems. If written in this way `verifyVersion` would always throw in v1 changelogs

Reply:
that's why I had overridden `readVersion` above in `StateStoreChangelogReaderV1`

Reply:
I ended up keeping this. Because `StateStoreChangelogReaderV3` extends `StateStoreChangelogReaderV1` (for it's iterator functions), now the readVersion is also inherited. We could override it by 
```
class StateStoreChangelogReaderV3 {
  override protected def readVersion(): String = input.readUTF()
}
```

Also future new readers extending this V1 again needs to override `input.readUTF()`. I'm thinking this defeats the purpose of overrding the readVersion override in V1, which is try to prevent unintentional overrides. 

Sadly because of this version difference it looks inevitable that there is no way to perfectly prevent unintentional overrides..